File: JsWs/apps/lan-streamer/api/src/audio/mixer.ts
import type { WebSocket } from "ws";
import { LIMITER_ATTACK_SECONDS, LIMITER_RELEASE_SECONDS } from "../config.js";
import { getAccountSettings } from "../store.js";

export const SAMPLE_RATE = 48000;
export const FRAME_SAMPLES = 960; // 20ms at 48kHz, per channel
export const CHANNELS = 2;
export const BYTES_PER_FRAME = FRAME_SAMPLES * CHANNELS * 2; // 16-bit PCM, interleaved [L0,R0,L1,R1,...]

const TICK_MS = 20;

// One-pole envelope smoothing coefficients for the per-device limiter: fast
// attack so a loud transient is caught almost immediately, slower release so
// gain eases back up smoothly afterward instead of snapping (which is what
// produces an audible "pop").
const LIMITER_ATTACK_COEFF = Math.exp(-1 / (SAMPLE_RATE * LIMITER_ATTACK_SECONDS));
const LIMITER_RELEASE_COEFF = Math.exp(-1 / (SAMPLE_RATE * LIMITER_RELEASE_SECONDS));

interface HostConnection {
  queue: Buffer[];
  accountId: number;
  /** current smoothed limiter gain (1 == no reduction), persists across frames/ticks */
  limiterGain: number;
  /** diagnostics: how many consecutive ticks this connection's queue has been empty */
  emptyStreak: number;
}

interface StreamMixState {
  // keyed per host *connection* (not per account) so that multiple windows/devices
  // under the same Filen account can stream into the same stream simultaneously
  // and have their audio mixed together rather than colliding on one queue.
  hostQueues: Map<string, HostConnection>;
  listeners: Set<WebSocket>;
  timer: NodeJS.Timeout;
  /** diagnostics: wall-clock time of the previous tick, to detect the mixer's own timer lagging */
  lastTickAt: number;
  // Reused every tick instead of allocating fresh arrays - with N host
  // connections, the old code allocated 1 Int32Array + N Int16Arrays *every
  // single 20ms tick* (50x/sec), which is enough garbage to trigger V8 GC
  // pauses landing the mixer's own setInterval noticeably behind schedule -
  // the actual cause of the periodic gaps this was built to diagnose.
  accumulator: Int32Array;
  // Also reused every tick (not just the accumulator) - a fresh
  // Buffer.alloc(BYTES_PER_FRAME) every tick was left in place under the
  // assumption that ws.send() copies synchronously before returning, but
  // that assumption was never actually verified, and tick-lag warnings kept
  // recurring even after the accumulator fix above. Since the *same* buffer
  // instance is already sent to every listener within one tick without
  // corruption, ws.send() provably does copy synchronously - so reusing it
  // across ticks too is just as safe as reusing the accumulator.
  output: Buffer;
}

const streams = new Map<string, StreamMixState>();

function ensureStream(streamId: string): StreamMixState {
  let state = streams.get(streamId);
  if (state) return state;

  state = {
    hostQueues: new Map(),
    listeners: new Set(),
    timer: setInterval(() => tick(streamId), TICK_MS),
    lastTickAt: Date.now(),
    accumulator: new Int32Array(FRAME_SAMPLES * CHANNELS),
    output: Buffer.alloc(BYTES_PER_FRAME),
  };
  streams.set(streamId, state);
  return state;
}

function maybeTeardown(streamId: string): void {
  const state = streams.get(streamId);
  if (!state) return;
  if (state.hostQueues.size === 0 && state.listeners.size === 0) {
    clearInterval(state.timer);
    streams.delete(streamId);
  }
}

/**
 * Limits a device stream's peak amplitude to `ceilingAmplitude` (fraction of
 * full-scale) and adds the result directly into `accumulator`, easing the
 * gain reduction in/out via an envelope follower rather than hard-clipping,
 * to avoid an audible "pop". Mutates `conn.limiterGain` to carry the
 * envelope across calls. Adds no buffering delay - the limiter acts on each
 * sample as it's read, it's just not a brick-wall clip.
 *
 * Stereo is "linked": both channels of a sample pair share one gain value
 * derived from whichever channel is louder, instead of being limited
 * independently. Independent per-channel limiting would shrink whichever
 * channel happens to be louder at any instant, smearing the stereo image
 * left/right as the signal moves - linking keeps the balance intact.
 */
function accumulateLimited(frame: Buffer, ceilingAmplitude: number, conn: HostConnection, accumulator: Int32Array): void {
  const ceiling = ceilingAmplitude * 32767;

  for (let i = 0; i < FRAME_SAMPLES; i++) {
    const left = frame.readInt16LE(i * CHANNELS * 2);
    const right = frame.readInt16LE((i * CHANNELS + 1) * 2);
    const magnitude = Math.max(Math.abs(left), Math.abs(right));
    const targetGain = magnitude > ceiling ? ceiling / magnitude : 1;
    const coeff = targetGain < conn.limiterGain ? LIMITER_ATTACK_COEFF : LIMITER_RELEASE_COEFF;
    conn.limiterGain = conn.limiterGain * coeff + targetGain * (1 - coeff);
    accumulator[i * CHANNELS] += Math.round(left * conn.limiterGain);
    accumulator[i * CHANNELS + 1] += Math.round(right * conn.limiterGain);
  }
}

interface SimpleStreamState {
  listeners: Set<WebSocket>;
  /** same envelope-follower limiter as accumulateLimited, but writing
   * directly into `output` instead of summing into an accumulator - a
   * simple stream only ever has one active source, so there's nothing to
   * mix and no clamping risk from multiple sources overlapping. */
  limiterGain: number;
  output: Buffer;
}

const simpleStreams = new Map<string, SimpleStreamState>();

function ensureSimpleStream(streamId: string): SimpleStreamState {
  let state = simpleStreams.get(streamId);
  if (state) return state;
  state = { listeners: new Set(), limiterGain: 1, output: Buffer.alloc(BYTES_PER_FRAME) };
  simpleStreams.set(streamId, state);
  return state;
}

function maybeTeardownSimple(streamId: string): void {
  const state = simpleStreams.get(streamId);
  if (state && state.listeners.size === 0) simpleStreams.delete(streamId);
}

/** Forwards one frame from the currently-active host straight to a simple
 * stream's listeners, applying the same account-wide volume cap as merged
 * streams but with no mixing/buffering tick involved - whatever the host
 * sends goes out immediately. */
export function forwardSimpleFrame(streamId: string, accountId: number, frame: Buffer): void {
  const state = simpleStreams.get(streamId);
  if (!state || state.listeners.size === 0) return;

  const { maxDeviceAmplitude } = getAccountSettings(accountId);
  const ceiling = maxDeviceAmplitude * 32767;
  for (let i = 0; i < FRAME_SAMPLES; i++) {
    const left = frame.readInt16LE(i * CHANNELS * 2);
    const right = frame.readInt16LE((i * CHANNELS + 1) * 2);
    const magnitude = Math.max(Math.abs(left), Math.abs(right));
    const targetGain = magnitude > ceiling ? ceiling / magnitude : 1;
    const coeff = targetGain < state.limiterGain ? LIMITER_ATTACK_COEFF : LIMITER_RELEASE_COEFF;
    state.limiterGain = state.limiterGain * coeff + targetGain * (1 - coeff);
    state.output.writeInt16LE(Math.round(left * state.limiterGain), i * CHANNELS * 2);
    state.output.writeInt16LE(Math.round(right * state.limiterGain), (i * CHANNELS + 1) * 2);
  }

  for (const ws of state.listeners) {
    if (ws.readyState === ws.OPEN) ws.send(state.output);
  }
}

export function registerSimpleListener(streamId: string, ws: WebSocket): void {
  ensureSimpleStream(streamId).listeners.add(ws);
}

export function unregisterSimpleListener(streamId: string, ws: WebSocket): void {
  const state = simpleStreams.get(streamId);
  if (!state) return;
  state.listeners.delete(ws);
  maybeTeardownSimple(streamId);
}

function tick(streamId: string): void {
  const state = streams.get(streamId);
  if (!state) return;

  // Diagnostics: if the mixer's own timer fires noticeably late, that points
  // at server-side load (event loop congestion, GC, etc.) rather than the
  // host failing to produce/send audio - distinguishes the two possible
  // sources of an output gap.
  const now = Date.now();
  const sinceLastTick = now - state.lastTickAt;
  state.lastTickAt = now;
  if (sinceLastTick > TICK_MS * 2) {
    console.warn(`[mixer] stream ${streamId}: tick fired ${sinceLastTick}ms after the previous one (expected ~${TICK_MS}ms) - mixer is running behind`);
  }

  if (state.listeners.size === 0) return;

  state.accumulator.fill(0);
  let anyContribution = false;

  for (const [connectionId, conn] of state.hostQueues) {
    const frame = conn.queue.shift();
    if (!frame) {
      conn.emptyStreak += 1;
      if (conn.emptyStreak === 1) {
        console.warn(`[mixer] stream ${streamId}: host ${connectionId} produced no frame for this tick (queue empty)`);
      }
      continue;
    }
    if (conn.emptyStreak > 1) {
      console.warn(`[mixer] stream ${streamId}: host ${connectionId} resumed after ${conn.emptyStreak * TICK_MS}ms with no frames`);
    }
    conn.emptyStreak = 0;
    anyContribution = true;

    const { maxDeviceAmplitude } = getAccountSettings(conn.accountId);
    accumulateLimited(frame, maxDeviceAmplitude, conn, state.accumulator);
  }

  if (!anyContribution) return;

  const mixed = state.output;
  for (let i = 0; i < state.accumulator.length; i++) {
    const clamped = Math.max(-32768, Math.min(32767, state.accumulator[i]));
    mixed.writeInt16LE(clamped, i * 2);
  }

  for (const ws of state.listeners) {
    if (ws.readyState === ws.OPEN) ws.send(mixed);
  }
}

export function registerHost(streamId: string, connectionId: string, accountId: number): void {
  const state = ensureStream(streamId);
  if (!state.hostQueues.has(connectionId)) {
    state.hostQueues.set(connectionId, { queue: [], accountId, limiterGain: 1, emptyStreak: 0 });
  }
}

export function unregisterHost(streamId: string, connectionId: string): void {
  const state = streams.get(streamId);
  if (!state) return;
  state.hostQueues.delete(connectionId);
  maybeTeardown(streamId);
}

export function pushHostFrame(streamId: string, connectionId: string, frame: Buffer): void {
  const state = streams.get(streamId);
  const conn = state?.hostQueues.get(connectionId);
  if (!conn) return;
  conn.queue.push(frame);
  // bound queue depth to avoid unbounded memory growth on slow consumers
  while (conn.queue.length > 25) conn.queue.shift();
}

export function registerListener(streamId: string, ws: WebSocket): void {
  ensureStream(streamId).listeners.add(ws);
}

export function unregisterListener(streamId: string, ws: WebSocket): void {
  const state = streams.get(streamId);
  if (!state) return;
  state.listeners.delete(ws);
  maybeTeardown(streamId);
}

File: JsWs/apps/lan-streamer/api-rs/src/mixer.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use tokio::sync::mpsc::UnboundedSender;
use tokio::task::AbortHandle;

use crate::config::{LIMITER_ATTACK_SECONDS, LIMITER_RELEASE_SECONDS};
use crate::store::Store;

pub const SAMPLE_RATE: u32 = 48000;
pub const FRAME_SAMPLES: usize = 960; // 20ms at 48kHz, per channel
pub const CHANNELS: usize = 2;
pub const BYTES_PER_FRAME: usize = FRAME_SAMPLES * CHANNELS * 2; // 16-bit PCM, interleaved [L0,R0,L1,R1,...]

const TICK_MS: u64 = 20;

/// One-pole envelope smoothing for the per-device limiter: fast attack so a
/// loud transient is caught almost immediately, slower release so gain eases
/// back up smoothly afterward instead of snapping (which is what produces an
/// audible "pop").
fn limiter_coeffs() -> (f64, f64) {
    let attack = (-1.0 / (SAMPLE_RATE as f64 * LIMITER_ATTACK_SECONDS)).exp();
    let release = (-1.0 / (SAMPLE_RATE as f64 * LIMITER_RELEASE_SECONDS)).exp();
    (attack, release)
}

/// One outgoing audio frame, sent by reference-counted clone to each
/// listener's forwarding channel rather than copied per-listener up front.
pub type Frame = Arc<[u8]>;

struct HostConnection {
    queue: Vec<Vec<u8>>,
    account_id: i64,
    /// current smoothed limiter gain (1 == no reduction), persists across frames/ticks
    limiter_gain: f64,
    /// diagnostics: how many consecutive ticks this connection's queue has been empty
    empty_streak: u32,
}

struct MergedStreamState {
    // keyed per host *connection* (not per account) so that multiple
    // windows/devices under the same Filen account can stream into the same
    // stream simultaneously and have their audio mixed together rather than
    // colliding on one queue.
    host_queues: HashMap<String, HostConnection>,
    listeners: HashMap<String, UnboundedSender<Frame>>,
    /// diagnostics: wall-clock time of the previous tick, to detect the mixer's own timer lagging
    last_tick_at: Instant,
    // Reused every tick instead of allocating fresh - matches the same fix
    // applied to the Node version (per-tick allocation was enough GC/alloc
    // pressure to land the tick timer noticeably behind schedule).
    accumulator: Vec<i32>,
    output: Vec<u8>,
    tick_task: AbortHandle,
}

struct SimpleStreamState {
    listeners: HashMap<String, UnboundedSender<Frame>>,
    limiter_gain: f64,
}

struct RawStreamState {
    listeners: HashMap<String, UnboundedSender<Frame>>,
}

pub struct ExclusiveHost {
    pub connection_id: String,
    pub close: tokio::sync::mpsc::UnboundedSender<()>,
}

pub struct Mixer {
    store: Arc<Store>,
    merged: Mutex<HashMap<String, MergedStreamState>>,
    simple: Mutex<HashMap<String, SimpleStreamState>>,
    raw: Mutex<HashMap<String, RawStreamState>>,
    pub active_simple_hosts: Mutex<HashMap<String, ExclusiveHost>>,
    pub active_raw_hosts: Mutex<HashMap<String, ExclusiveHost>>,
    limiter_attack_coeff: f64,
    limiter_release_coeff: f64,
}

impl Mixer {
    pub fn new(store: Arc<Store>) -> Self {
        let (attack, release) = limiter_coeffs();
        Self {
            store,
            merged: Mutex::new(HashMap::new()),
            simple: Mutex::new(HashMap::new()),
            raw: Mutex::new(HashMap::new()),
            active_simple_hosts: Mutex::new(HashMap::new()),
            active_raw_hosts: Mutex::new(HashMap::new()),
            limiter_attack_coeff: attack,
            limiter_release_coeff: release,
        }
    }

    // ---- merged ----

    pub fn register_host(self: &Arc<Self>, stream_id: &str, connection_id: &str, account_id: i64) {
        let mut merged = self.merged.lock().unwrap();
        let state = ensure_merged(self, &mut merged, stream_id);
        state.host_queues.entry(connection_id.to_string()).or_insert_with(|| HostConnection {
            queue: Vec::new(),
            account_id,
            limiter_gain: 1.0,
            empty_streak: 0,
        });
    }

    pub fn unregister_host(&self, stream_id: &str, connection_id: &str) {
        let mut merged = self.merged.lock().unwrap();
        if let Some(state) = merged.get_mut(stream_id) {
            state.host_queues.remove(connection_id);
            maybe_teardown_merged(&mut merged, stream_id);
        }
    }

    pub fn push_host_frame(&self, stream_id: &str, connection_id: &str, frame: Vec<u8>) {
        let mut merged = self.merged.lock().unwrap();
        if let Some(state) = merged.get_mut(stream_id) {
            if let Some(conn) = state.host_queues.get_mut(connection_id) {
                conn.queue.push(frame);
                // bound queue depth to avoid unbounded memory growth on slow consumers
                while conn.queue.len() > 25 {
                    conn.queue.remove(0);
                }
            }
        }
    }

    pub fn register_listener(self: &Arc<Self>, stream_id: &str, connection_id: String, sender: UnboundedSender<Frame>) {
        let mut merged = self.merged.lock().unwrap();
        let state = ensure_merged(self, &mut merged, stream_id);
        state.listeners.insert(connection_id, sender);
    }

    pub fn unregister_listener(&self, stream_id: &str, connection_id: &str) {
        let mut merged = self.merged.lock().unwrap();
        if let Some(state) = merged.get_mut(stream_id) {
            state.listeners.remove(connection_id);
            maybe_teardown_merged(&mut merged, stream_id);
        }
    }

    fn tick(&self, stream_id: &str) {
        let mut merged = self.merged.lock().unwrap();
        let Some(state) = merged.get_mut(stream_id) else { return };

        // Diagnostics: if the mixer's own timer fires noticeably late, that
        // points at server-side load (scheduler congestion, etc.) rather
        // than the host failing to produce/send audio.
        let now = Instant::now();
        let since_last_tick = now.duration_since(state.last_tick_at).as_millis() as u64;
        state.last_tick_at = now;
        if since_last_tick > TICK_MS * 2 {
            eprintln!("[mixer] stream {stream_id}: tick fired {since_last_tick}ms after the previous one (expected ~{TICK_MS}ms) - mixer is running behind");
        }

        if state.listeners.is_empty() {
            return;
        }

        state.accumulator.fill(0);
        let mut any_contribution = false;

        for (connection_id, conn) in state.host_queues.iter_mut() {
            let Some(frame) = (!conn.queue.is_empty()).then(|| conn.queue.remove(0)) else {
                conn.empty_streak += 1;
                if conn.empty_streak == 1 {
                    eprintln!("[mixer] stream {stream_id}: host {connection_id} produced no frame for this tick (queue empty)");
                }
                continue;
            };
            if conn.empty_streak > 1 {
                eprintln!(
                    "[mixer] stream {stream_id}: host {connection_id} resumed after {}ms with no frames",
                    conn.empty_streak as u64 * TICK_MS
                );
            }
            conn.empty_streak = 0;
            any_contribution = true;

            let settings = self.store.get_account_settings(conn.account_id);
            accumulate_limited(&frame, settings.max_device_amplitude, conn, &mut state.accumulator, self.limiter_attack_coeff, self.limiter_release_coeff);
        }

        if !any_contribution {
            return;
        }

        for (i, sample) in state.accumulator.iter().enumerate() {
            let clamped = (*sample).clamp(-32768, 32767) as i16;
            state.output[i * 2..i * 2 + 2].copy_from_slice(&clamped.to_le_bytes());
        }

        let frame: Frame = Arc::from(state.output.as_slice());
        state.listeners.retain(|_, sender| sender.send(frame.clone()).is_ok());
    }

    // ---- simple ----

    pub fn register_simple_listener(&self, stream_id: &str, connection_id: String, sender: UnboundedSender<Frame>) {
        let mut simple = self.simple.lock().unwrap();
        simple.entry(stream_id.to_string()).or_insert_with(|| SimpleStreamState { listeners: HashMap::new(), limiter_gain: 1.0 }).listeners.insert(connection_id, sender);
    }

    pub fn unregister_simple_listener(&self, stream_id: &str, connection_id: &str) {
        let mut simple = self.simple.lock().unwrap();
        if let Some(state) = simple.get_mut(stream_id) {
            state.listeners.remove(connection_id);
            if state.listeners.is_empty() {
                simple.remove(stream_id);
            }
        }
    }

    /// Forwards one frame from the currently-active host straight to a
    /// simple stream's listeners, applying the same account-wide volume cap
    /// as merged streams but with no mixing/buffering tick involved.
    pub fn forward_simple_frame(&self, stream_id: &str, account_id: i64, frame: &[u8]) {
        let mut simple = self.simple.lock().unwrap();
        let Some(state) = simple.get_mut(stream_id) else { return };
        if state.listeners.is_empty() {
            return;
        }

        let settings = self.store.get_account_settings(account_id);
        let ceiling = settings.max_device_amplitude * 32767.0;
        let mut output = vec![0u8; BYTES_PER_FRAME];
        for i in 0..FRAME_SAMPLES {
            let left = i16::from_le_bytes([frame[i * CHANNELS * 2], frame[i * CHANNELS * 2 + 1]]) as f64;
            let right = i16::from_le_bytes([frame[(i * CHANNELS + 1) * 2], frame[(i * CHANNELS + 1) * 2 + 1]]) as f64;
            let magnitude = left.abs().max(right.abs());
            let target_gain = if magnitude > ceiling { ceiling / magnitude } else { 1.0 };
            let coeff = if target_gain < state.limiter_gain { self.limiter_attack_coeff } else { self.limiter_release_coeff };
            state.limiter_gain = state.limiter_gain * coeff + target_gain * (1.0 - coeff);
            let l = (left * state.limiter_gain).round() as i16;
            let r = (right * state.limiter_gain).round() as i16;
            output[i * CHANNELS * 2..i * CHANNELS * 2 + 2].copy_from_slice(&l.to_le_bytes());
            output[(i * CHANNELS + 1) * 2..(i * CHANNELS + 1) * 2 + 2].copy_from_slice(&r.to_le_bytes());
        }

        let frame: Frame = Arc::from(output.as_slice());
        state.listeners.retain(|_, sender| sender.send(frame.clone()).is_ok());
    }

    // ---- raw ----

    pub fn register_raw_listener(&self, stream_id: &str, connection_id: String, sender: UnboundedSender<Frame>) {
        let mut raw = self.raw.lock().unwrap();
        raw.entry(stream_id.to_string()).or_insert_with(|| RawStreamState { listeners: HashMap::new() }).listeners.insert(connection_id, sender);
    }

    pub fn unregister_raw_listener(&self, stream_id: &str, connection_id: &str) {
        let mut raw = self.raw.lock().unwrap();
        if let Some(state) = raw.get_mut(stream_id) {
            state.listeners.remove(connection_id);
            if state.listeners.is_empty() {
                raw.remove(stream_id);
            }
        }
    }

    /// Forwards a frame completely unprocessed - no volume cap, no limiter.
    /// For anyone who wants bit-exact passthrough with none of the safety
    /// processing the other two modes apply.
    pub fn forward_raw_frame(&self, stream_id: &str, frame: &[u8]) {
        let raw = self.raw.lock().unwrap();
        let Some(state) = raw.get(stream_id) else { return };
        if state.listeners.is_empty() {
            return;
        }
        let frame: Frame = Arc::from(frame);
        for sender in state.listeners.values() {
            let _ = sender.send(frame.clone());
        }
    }
}

/// Limits a device stream's peak amplitude to `ceiling_amplitude` (fraction
/// of full-scale) and adds the result directly into `accumulator`, easing
/// the gain reduction in/out via an envelope follower rather than
/// hard-clipping, to avoid an audible "pop".
///
/// Stereo is "linked": both channels of a sample pair share one gain value
/// derived from whichever channel is louder, instead of being limited
/// independently - independent per-channel limiting would shrink whichever
/// channel happens to be louder at any instant, smearing the stereo image
/// left/right as the signal moves.
fn accumulate_limited(frame: &[u8], ceiling_amplitude: f64, conn: &mut HostConnection, accumulator: &mut [i32], attack_coeff: f64, release_coeff: f64) {
    let ceiling = ceiling_amplitude * 32767.0;
    for i in 0..FRAME_SAMPLES {
        let left = i16::from_le_bytes([frame[i * CHANNELS * 2], frame[i * CHANNELS * 2 + 1]]) as f64;
        let right = i16::from_le_bytes([frame[(i * CHANNELS + 1) * 2], frame[(i * CHANNELS + 1) * 2 + 1]]) as f64;
        let magnitude = left.abs().max(right.abs());
        let target_gain = if magnitude > ceiling { ceiling / magnitude } else { 1.0 };
        let coeff = if target_gain < conn.limiter_gain { attack_coeff } else { release_coeff };
        conn.limiter_gain = conn.limiter_gain * coeff + target_gain * (1.0 - coeff);
        accumulator[i * CHANNELS] += (left * conn.limiter_gain).round() as i32;
        accumulator[i * CHANNELS + 1] += (right * conn.limiter_gain).round() as i32;
    }
}

fn ensure_merged<'a>(mixer: &Arc<Mixer>, merged: &'a mut HashMap<String, MergedStreamState>, stream_id: &str) -> &'a mut MergedStreamState {
    if !merged.contains_key(stream_id) {
        let mixer = mixer.clone();
        let id = stream_id.to_string();
        let handle = tokio::spawn(async move {
            let mut interval = tokio::time::interval(std::time::Duration::from_millis(TICK_MS));
            loop {
                interval.tick().await;
                mixer.tick(&id);
            }
        });
        merged.insert(
            stream_id.to_string(),
            MergedStreamState {
                host_queues: HashMap::new(),
                listeners: HashMap::new(),
                last_tick_at: Instant::now(),
                accumulator: vec![0i32; FRAME_SAMPLES * CHANNELS],
                output: vec![0u8; BYTES_PER_FRAME],
                tick_task: handle.abort_handle(),
            },
        );
    }
    merged.get_mut(stream_id).unwrap()
}

fn maybe_teardown_merged(merged: &mut HashMap<String, MergedStreamState>, stream_id: &str) {
    if let Some(state) = merged.get(stream_id) {
        if state.host_queues.is_empty() && state.listeners.is_empty() {
            state.tick_task.abort();
            merged.remove(stream_id);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::AccountSettings;
    use std::time::Duration;

    fn make_frame(left: i16, right: i16) -> Vec<u8> {
        let mut buf = vec![0u8; BYTES_PER_FRAME];
        for i in 0..FRAME_SAMPLES {
            buf[i * CHANNELS * 2..i * CHANNELS * 2 + 2].copy_from_slice(&left.to_le_bytes());
            buf[(i * CHANNELS + 1) * 2..(i * CHANNELS + 1) * 2 + 2].copy_from_slice(&right.to_le_bytes());
        }
        buf
    }

    fn read_sample(frame: &[u8], index: usize) -> i16 {
        i16::from_le_bytes([frame[index * 2], frame[index * 2 + 1]])
    }

    fn test_store() -> Arc<Store> {
        Arc::new(Store::new(std::env::temp_dir().join(format!("lan-streamer-test-{}", uuid::Uuid::new_v4()))))
    }

    #[tokio::test]
    async fn merged_sums_two_hosts_per_channel() {
        let store = test_store();
        store.set_account_settings(1, AccountSettings { max_device_amplitude: 1.0 }).await;
        let mixer = Arc::new(Mixer::new(store));
        let stream_id = "regress-merged-sum";

        mixer.register_host(stream_id, "conn-a", 1);
        mixer.register_host(stream_id, "conn-b", 1);
        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
        mixer.register_listener(stream_id, "listener-1".into(), tx);

        mixer.push_host_frame(stream_id, "conn-a", make_frame(1000, 2000));
        mixer.push_host_frame(stream_id, "conn-b", make_frame(3000, -500));

        let frame = tokio::time::timeout(Duration::from_millis(500), rx.recv()).await.expect("timed out").expect("channel closed");
        assert_eq!(read_sample(&frame, 0), 4000, "left channels should sum");
        assert_eq!(read_sample(&frame, 1), 1500, "right channels should sum");

        mixer.unregister_host(stream_id, "conn-a");
        mixer.unregister_host(stream_id, "conn-b");
        mixer.unregister_listener(stream_id, "listener-1");
    }

    #[tokio::test]
    async fn merged_limiter_preserves_stereo_ratio_while_capping() {
        let store = test_store();
        store.set_account_settings(2, AccountSettings { max_device_amplitude: 0.5 }).await;
        let mixer = Arc::new(Mixer::new(store));
        let stream_id = "regress-merged-limiter";

        mixer.register_host(stream_id, "conn-a", 2);
        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
        mixer.register_listener(stream_id, "listener-1".into(), tx);

        // Full-scale left, 10%-amplitude right - ratio must survive limiting.
        for _ in 0..50 {
            mixer.push_host_frame(stream_id, "conn-a", make_frame(32767, 3277));
            let frame = tokio::time::timeout(Duration::from_millis(200), rx.recv()).await.expect("timed out").expect("channel closed");
            let l = read_sample(&frame, 0) as f64;
            let r = read_sample(&frame, 1) as f64;
            if l.abs() > 1.0 {
                let ratio = r / l;
                assert!((ratio - 0.1).abs() < 0.01, "L/R ratio should stay ~0.1, got {ratio}");
            }
        }

        mixer.unregister_host(stream_id, "conn-a");
        mixer.unregister_listener(stream_id, "listener-1");
    }

    #[tokio::test]
    async fn simple_forward_applies_cap_but_raw_does_not() {
        let store = test_store();
        store.set_account_settings(3, AccountSettings { max_device_amplitude: 0.1 }).await;
        let mixer = Mixer::new(store);

        let (tx_simple, mut rx_simple) = tokio::sync::mpsc::unbounded_channel();
        mixer.register_simple_listener("regress-simple", "l1".into(), tx_simple);
        mixer.forward_simple_frame("regress-simple", 3, &make_frame(32000, -32000));
        let simple_frame = rx_simple.recv().await.expect("no frame");
        let capped = read_sample(&simple_frame, 0);
        assert!(capped.abs() < 32000, "simple mode should cap toward the 10% ceiling, got {capped}");

        let (tx_raw, mut rx_raw) = tokio::sync::mpsc::unbounded_channel();
        mixer.register_raw_listener("regress-raw", "l1".into(), tx_raw);
        mixer.forward_raw_frame("regress-raw", &make_frame(32000, -32000));
        let raw_frame = rx_raw.recv().await.expect("no frame");
        assert_eq!(read_sample(&raw_frame, 0), 32000, "raw mode must bypass the cap entirely");
        assert_eq!(read_sample(&raw_frame, 1), -32000);
    }
}
